todo_app/app.py

```diff
@@ -62,10 +161,42 @@ def change_todo_status(id):
         return redirect('/')
 
     @app.route('/todo/delete/<id>', methods=['POST'])
+    @login_required
+    @User.check_permission('write')
     def delete_todo(id):
         db = DB()
         item_to_delete = db.get_item(id)
         db.delete_item(item_to_delete)
         return redirect('/')
+
+    @app.route('/login/error', methods=['GET'])
+    def login_error():
+        return render_template('login_error.html')            
+
+    @app.route('/user/management', methods=['GET'])
+    @User.check_permission('admin')
+    def user_management():
+
+        user_view_model = UserViewModel(current_user)
+        user_management = UserManagement()
+        users = user_management.get_users()
+
+        return render_template('user_management.html',
+            user_view_model = user_view_model,
+            users = users,
+            roles = Roles
+        )
+
+    @User.check_permission('admin')
+    @app.route('/user/management/update/<id>', methods=['POST'])
+    def update_user(id):
```

Comment:
Interesting, I'll take a look when I get the chance. 

I did revert the UI changes to how they were after I had implemented this and I wasn't able to perform writes/updates. As a reader. I guess I forgot to specifically check for the users page.

Reply:
I do actually remember reading the warning about this in the instructions which makes it more annoying I missed it.

Reply:
It's hard to test all the combos, I missed this originally in my own direct testing too and it's only because I've seen people bitten by it a few times I spotted it in the code review!

Reply:
I have now updated this. It makes sense why it wasn't working correctly.

From my understanding, the `@User.check_permission` decorator would get called once on startup before flask registers the route? Requests for that endpoint just call the `update_user` function without the permission check.